In [19]:
import pandas as pd
import numpy as np
import requests
import spacy
import folium
import matplotlib.pyplot as plt
import seaborn as sns
from folium import plugins
from funciones_analisis import analizar_dataframe
from funciones_analisis import convertir_notacion_india_numero_occidental

In [17]:
df=pd.read_excel(r"E:\Cursos\BC_Data_Science\Repositorio\ONLINE_DS_THEBRIDGE_V\proyecto EDA\DATASETS\DATASETS MODIFICADOS\Ciudades_india_modificado.xlsx")

In [18]:
df

Ciudades      Estados                      Tipo_area_urbana  \
0     Bengaluru    Karnataka   Municipal Corporation / Corporation   
1     Ahmedabad      Gujarat  Municipal Corporation / Corporation.   
2     Hyderabad    Telangana   Municipal Corporation / Corporation   
3       Chennai   Tamil Nadu  Municipal Corporation / Corporation.   
4       Kolkata  West Bengal   Municipal Corporation / Corporation   
...         ...          ...                                   ...   
1210        NaN          NaN                                   NaN   
1211        NaN          NaN                                   NaN   
1212        NaN          NaN                                   NaN   
1213        NaN          NaN                                   NaN   
1214        NaN          NaN                                   NaN   

     Clases_sociales  Numero_Occidental              Tipo_ciudad  
0            Class I        842500970.0  Ciudades_metropolitanas  
1            Class I        720800200.0  Ciudades_metropolitanas  
2            Class I        680900970.0  Ciudades_metropolitanas  
3            Class I        468100087.0  Ciudades_metropolitanas  
4            Class I        448600679.0  Ciudades_metropolitanas  
...              ...                ...                      ...  
1210             NaN                NaN         Ciudades_menores  
1211             NaN                NaN         Ciudades_menores  
1212             NaN                NaN         Ciudades_menores  
1213             NaN                NaN         Ciudades_menores  
1214             NaN                NaN         Ciudades_menores  

[1215 rows x 6 columns]

In [16]:
df.to_excel(r"E:\Cursos\BC_Data_Science\Repositorio\ONLINE_DS_THEBRIDGE_V\proyecto EDA\DATASETS\DATASETS MODIFICADOS\Ciudades_india_modificado.xlsx", index = False)

In [5]:
df['Numero_Occidental'] = df["Population (2011)"].apply(convertir_notacion_india_numero_occidental)

In [6]:
df_ordenado= df.sort_values(by="Numero_Occidental", ascending=False, inplace =True)

In [7]:
def ciudades_india(numero):
 
    if numero > 10000000:
         return "Ciudades_metropolitanas"
    elif numero >1000000:
        return "Ciudades_principales"
    else:
        return "Ciudades_menores"

In [8]:
df["Tipo_ciudad"] = df['Numero_Occidental'].apply(ciudades_india)

In [65]:
df.rename(columns={"Name of City": "Ciudades"}, inplace=True)

In [20]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1215 entries, 0 to 1214
Data columns (total 6 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Ciudades           1209 non-null   object 
 1   Estados            1209 non-null   object 
 2   Tipo_area_urbana   1206 non-null   object 
 3   Clases_sociales    661 non-null    object 
 4   Numero_Occidental  1209 non-null   float64
 5   Tipo_ciudad        1215 non-null   object 
dtypes: float64(1), object(5)
memory usage: 57.1+ KB


In [6]:
#analizo el DF
resumen_analisis = analizar_dataframe(df)
resumen_analisis

Tipo  Cardinalidad  Porcentaje_Nulos Clasificación
Ciudades           object          1190          0.493827    Categórica
Estados            object            34          0.493827    Categórica
Tipo_area_urbana   object            34          0.740741    Categórica
Population (2011)  object          1193          0.493827    Categórica
Clases_sociales    object             5         45.596708    Categórica

In [22]:
#me traigo mi dataframe , para sacar la lista de ciudades origen de los ataques con este dataframe
df_eda=pd.read_excel(r"E:\Cursos\BC_Data_Science\Repositorio\ONLINE_DS_THEBRIDGE_V\proyecto EDA\DATASETS\DATASETS MODIFICADOS\ciber_modificado_ciudades_final.xlsx")

In [8]:
df_eda.head(5)

Anyo      Fecha      Hora       IP Origen       Destino IP    Ciudades  \
0  2023 2023-05-30  06:33:58   103.216.15.12     84.9.164.252  Jamshedpur   
1  2020 2020-08-26  07:08:30  78.199.217.198   66.191.137.154    Bilaspur   
2  2022 2022-11-13  08:23:25    63.79.210.48    198.219.82.17      Bokaro   
3  2023 2023-07-02  10:38:46   163.42.196.10  101.228.192.255     Jaunpur   
4  2023 2023-07-16  13:11:07   71.166.185.76  189.243.174.238   Anantapur   

     Estados Tipo_ataque Nivel_severidad  \
0     Sikkim     Malware             Low   
1   Nagaland     Malware             Low   
2  Rajasthan        DDoS             Low   
3  Rajasthan     Malware          Medium   
4    Tripura        DDoS             Low   

                                    Datos_carga_util  \
0  Qui natus odio asperiores nam. Optio nobis ius...   
1  Aperiam quos modi officiis veritatis rem. Omni...   
2  Perferendis sapiente vitae soluta. Hic delectu...   
3  Totam maxime beatae expedita explicabo porro l...   
4  Odit nesciunt dolorem nisi iste iusto. Animi v...   

                                  Resultado_analisis Etiqueta_x  \
0  {'entidades': [], 'partes_del_discurso': ['PRO...     normal   
1  {'entidades': ['Aperiam', 'Illo', 'Quidem', 'C...    anómala   
2  {'entidades': ['quae'], 'partes_del_discurso':...    anómala   
3  {'entidades': ['porro labore', 'Minima', 'fugi...     normal   
4  {'entidades': ['nisi', 'Animi', 'illo repellen...     normal   

  Informacion_usuario                            Informacion_dispositivo  \
0       Reyansh Dugal  Mozilla/5.0 (compatible; MSIE 8.0; Windows NT ...   
1          Sumer Rana  Mozilla/5.0 (compatible; MSIE 8.0; Windows NT ...   
2        Himmat Karpe  Mozilla/5.0 (compatible; MSIE 9.0; Windows NT ...   
3          Fateh Kibe  Mozilla/5.0 (Macintosh; PPC Mac OS X 10_11_5; ...   
4        Dhanush Chad  Mozilla/5.0 (compatible; MSIE 5.0; Windows NT ...   

  Indicadores_malware  Puntuaciones_anomalias  
0        IoC Detected                    2867  
1        IoC Detected                     515  
2        IoC Detected                    8742  
3                 NaN                    1579  
4                 NaN                      52

In [32]:
print(len(df_eda.Estados.unique()))

28


In [33]:
print(len(df_eda.Ciudades.unique()))

316


In [45]:
#creamo tro nuevo dataframe modificado con ciudad y estado separado
columnas_elegidas = ["Anyo", "Fecha", "Hora", "IP Origen", "Destino IP","Ciudades","Estados","Tipo_ataque","Nivel_severidad","Datos_carga_util","Resultado_analisis","Etiqueta_x","Informacion_usuario","Informacion_dispositivo", "Indicadores_malware", "Puntuaciones_anomalias"]

# Crear un nuevo DataFrame con las columnas seleccionadas y en el orden deseado
ciber_modificado_ciudades = df_eda[columnas_elegidas].copy()

In [47]:
ciber_modificado_ciudades.to_excel(r"E:\Cursos\BC_Data_Science\Repositorio\ONLINE_DS_THEBRIDGE_V\proyecto EDA\DATASETS\DATASETS MODIFICADOS\ciber_modificado_ciudades_final.xlsx", index=False)


In [10]:
df["Clases_sociales"].value_counts()

Clases_sociales
Class I      254
Class III    205
Class II     183
Class IV      12
Class V        7
Name: count, dtype: int64

In [27]:

# Contar las ocurrencias de cada combinación de 'Ciudades' y 'Estados' en ambos DataFrames
num_df_eda = df_eda.groupby(['Ciudades', 'Estados']).size().reset_index(name='Conteo_eda')# agrupamos por ciudades y estados contando con size por valores unicos, y agrupando el resultado
# en la columna dela variable name
num_df = df.groupby(['Ciudades', 'Estados']).size().reset_index(name='Conteo_df')

# Fusionar los resultados en un nuevo DataFrame
resultado = pd.merge(num_df_eda, num_df, on=['Ciudades', 'Estados'], how='outer').fillna(0)

# Calcular la suma total de ocurrencias
resultado['Total'] = resultado['Conteo_eda'] + resultado['Conteo_df']

# Filtrar las filas donde hay coincidencias (donde Total > 0)
df_conteo_ciudades = resultado[resultado['Total'] > 0]

# Mostrar el resultado
df_conteo_ciudades.head(1)


Ciudades         Estados  Conteo_eda  Conteo_df  Total
0    Adoni  Andhra Pradesh         4.0        1.0    5.0

In [29]:
print(len(num_df_eda))

8723


In [30]:
print(len(num_df))

1205


In [26]:
resultado = df_conteo_ciudades.loc[(df_conteo_ciudades["Conteo_eda"] >= 1) & (df_conteo_ciudades["Conteo_df"] >= 1)]

print(len(resultado))
resultado

197


Ciudades         Estados  Conteo_eda  Conteo_df  Total
0             Adoni  Andhra Pradesh         4.0        1.0    5.0
52         Agartala         Tripura         3.0        1.0    4.0
81             Agra   Uttar Pradesh         6.0        1.0    7.0
90        Ahmedabad         Gujarat         5.0        1.0    6.0
125      Ahmednagar     Maharashtra         4.0        1.0    5.0
...             ...             ...         ...        ...    ...
8578        Vellore      Tamil Nadu         5.0        1.0    6.0
8612     Vijayawada  Andhra Pradesh         2.0        1.0    3.0
8640  Visakhapatnam  Andhra Pradesh         7.0        1.0    8.0
8690       Warangal       Telangana         6.0        1.0    7.0
8702    Yamunanagar         Haryana         2.0        1.0    3.0

[197 rows x 5 columns]

In [23]:
df_eda.head(1)

Anyo      Fecha      Hora      IP Origen    Destino IP    Ciudades Estados  \
0  2023 2023-05-30  06:33:58  103.216.15.12  84.9.164.252  Jamshedpur  Sikkim   

  Tipo_ataque Nivel_severidad  \
0     Malware             Low   

                                    Datos_carga_util  \
0  Qui natus odio asperiores nam. Optio nobis ius...   

                                  Resultado_analisis Etiqueta_x  \
0  {'entidades': [], 'partes_del_discurso': ['PRO...     normal   

  Informacion_usuario                            Informacion_dispositivo  \
0       Reyansh Dugal  Mozilla/5.0 (compatible; MSIE 8.0; Windows NT ...   

  Indicadores_malware  Puntuaciones_anomalias  
0        IoC Detected                    2867

In [26]:
df

Ciudades      Estados                      Tipo_area_urbana  \
0     Bengaluru    Karnataka   Municipal Corporation / Corporation   
1     Ahmedabad      Gujarat  Municipal Corporation / Corporation.   
2     Hyderabad    Telangana   Municipal Corporation / Corporation   
3       Chennai   Tamil Nadu  Municipal Corporation / Corporation.   
4       Kolkata  West Bengal   Municipal Corporation / Corporation   
...         ...          ...                                   ...   
1210        NaN          NaN                                   NaN   
1211        NaN          NaN                                   NaN   
1212        NaN          NaN                                   NaN   
1213        NaN          NaN                                   NaN   
1214        NaN          NaN                                   NaN   

     Clases_sociales  Numero_Occidental              Tipo_ciudad  
0            Class I        842500970.0  Ciudades_metropolitanas  
1            Class I        720800200.0  Ciudades_metropolitanas  
2            Class I        680900970.0  Ciudades_metropolitanas  
3            Class I        468100087.0  Ciudades_metropolitanas  
4            Class I        448600679.0  Ciudades_metropolitanas  
...              ...                ...                      ...  
1210             NaN                NaN         Ciudades_menores  
1211             NaN                NaN         Ciudades_menores  
1212             NaN                NaN         Ciudades_menores  
1213             NaN                NaN         Ciudades_menores  
1214             NaN                NaN         Ciudades_menores  

[1215 rows x 6 columns]

In [32]:

# Crear un DataFrame con la comparación de 'Ciudades' y 'Estados'
resultado_2 = pd.merge(df[['Ciudades', 'Estados', 'Tipo_ciudad']], df_eda[['Ciudades', 'Estados']], on=['Ciudades', 'Estados'], how='outer', suffixes=('_df1', '_df2'))

# Mostrar el resultado
resultado_2


Ciudades        Estados              Tipo_ciudad
0      Bengaluru      Karnataka  Ciudades_metropolitanas
1      Ahmedabad        Gujarat  Ciudades_metropolitanas
2      Ahmedabad        Gujarat  Ciudades_metropolitanas
3      Ahmedabad        Gujarat  Ciudades_metropolitanas
4      Ahmedabad        Gujarat  Ciudades_metropolitanas
...          ...            ...                      ...
41013     Purnia  Uttar Pradesh                      NaN
41014    Cuttack         Sikkim                      NaN
41015   Vadodara      Meghalaya                      NaN
41016    Cuttack      Jharkhand                      NaN
41017     Nagpur         Punjab                      NaN

[41018 rows x 3 columns]

In [31]:
resultado_2.value_counts()

Ciudades      Estados        Tipo_ciudad            
Aurangabad    Bihar          Ciudades_metropolitanas    10
Hyderabad     Telangana      Ciudades_metropolitanas    10
Munger        Bihar          Ciudades_metropolitanas    10
Jodhpur       Rajasthan      Ciudades_metropolitanas    10
Meerut        Uttar Pradesh  Ciudades_metropolitanas     9
                                                        ..
Nadbai        Rajasthan      Ciudades_principales        1
Nagaon        Assam          Ciudades_metropolitanas     1
Nagapattinam  Tamil Nadu     Ciudades_metropolitanas     1
Nagar         Rajasthan      Ciudades_principales        1
Zunheboto     Nagaland       Ciudades_principales        1
Name: count, Length: 1207, dtype: int64

In [33]:
ciudades_importantes= ["Delhi", "Mumbai","Kolkata","Chennai","Bengaluru","Hyderabad","Ahmedabad","Pune","Surat" ]

# Filtrar el DataFrame resultado para incluir solo las ciudades de interés
ciudades_interes = resultado_2[resultado_2['Ciudades'].isin(ciudades_importantes)]

# Obtener el conteo de cada ciudad de interés
conteo_ciudades_interes = ciudades_interes['Ciudades'].value_counts()


conteo_ciudades_interes


Ciudades
Hyderabad    129
Ahmedabad    125
Delhi        121
Mumbai       121
Kolkata      116
Chennai      115
Pune         112
Surat        109
Bengaluru      1
Name: count, dtype: int64

La India tien un total de 28 Estados y un total de 7908 ciudades en total. En mi df_eda aparecen los 28 estados, y 1209 ciudades, es decir ciudades de todos los estados. Ademas las ciudades mas importantes de la India son Delhi, Mumbai,Kolkata , Chennai , Bengaluru , Hyderabad , Ahmedabad , Pune y Surat , aparecen repetidas un numero muy elevado de veces, euivalentes a numero de ataques de origen o destino de  las 10 ciudades mas importantes de la india.